parallel operations for some binary caching providers

Crzyrndm opened the pull request:
https://github.com/microsoft/vcpkg/issues/38404

This adds parallel operations using `parallel_for_each` /  `parallel_transform` to `ObjectStorageProvider`. There is currently no limitation placed on the parallelisation past that imposed by the threadpool.

This impacts the following caching providers
* `GcsStorageTool`
* `AwsStorageTool`
* `CosStorageTool`

Questions
- [ ] This adds parallelisation at the provider level. Is this the right place to add it?
- [ ] Testing?
  * I couldn't find any tests for the affected caching providers and this isn't something I would normally test for anyway (implementation detail).

Crzyrndm committed:
parallel operations for some binary caching providers

#38404

Crzyrndm commented:
@microsoft-github-policy-service agree

BillyONeal commented:
In general, you're parallelizing the download here, but we might expect most of the time this to not be CPU limited and thus not improve as a result of parallelization. Do you have a real world use case where this did end up being meaningfully faster?

Crzyrndm commented:
For clarity - my experience here is with an AWS S3 cache specifically
I fully agree that this is an IO limited task and therefore threading isn't exactly the best model. This was more a quick demo to show there were significant potential improvements

Taking a project I have with 10 small dependencies (download size is minimal)
`> $Env:VCPKG_BINARY_SOURCES="clear;x-aws,s3://<test bucket name>/,read"`

## AWS S3 cache miss

Using the default vcpkg release binary - querying for presence takes 9.5s
![image](https://github.com/microsoft/vcpkg-tool/assets/8765278/8b51114e-2ad7-4e59-af1b-b8bb88c5bcc1)
Using the custom vcpkg binary from this branch - querying for presence takes 1.7s
![image](https://github.com/microsoft/vcpkg-tool/assets/8765278/3870a462-6156-4c38-b1c8-3a8dc7138d3a)

Note that typically I only use the cache for CI (GHA) only and the time per package can be 1-3 seconds. The bucket region may play a significant part in that but I havn't got around to testing yet

## AWS S3 cache hit

Using the default release binary - query and download takes 23s
![image](https://github.com/microsoft/vcpkg-tool/assets/8765278/fd8debac-ec3a-4160-b4f0-faf814eb3ce3)
Using the custom vcpkg binary - query and download takes 4.3s
![image](https://github.com/microsoft/vcpkg-tool/assets/8765278/ddf45486-3b35-45d1-90f6-39348eecae3e)

So approx  a 5x improvement for both query and download and that should be fairly proportional to the total number of dependencies (if this was done correctly). Writing back doesn't scale the same way because it's done after each package is built so we just ignore that for the moment

BillyONeal commented:
I agree a real world 5x improvement is worth opening a can of worms for; OK

Crzyrndm committed:
revert changes to upload path

this is pushing to the different caches(?) in parallel, not by package
like the others

Crzyrndm commented:
Thoughts on where to go with this?
If you've got an example of parallelisation of shell commands that would be more appropriate than the quick and dirty `parallel_*` I used here I could take a whack at that. The data flow does lend itself to parallelisation well and the CLI tools for AWS/GCP are setup for multi-process operations to my knowledge so the correctness _probably_ isn't as much of an issue as it could be (famous last words...)

Crzyrndm committed:
remove the variable rename

Crzyrndm committed:
add comment noting calls will occur in parallel

Crzyrndm commented:
`cmd_execute_and_capture_output_parallel` is doing pretty much the same thing I'm doing here (would have to significantly reshuffle the AWS/GCP/etc. API to use that directly. The minimal form is to use the parallel functions to replace the loop as has been done.

* Added a comment to the `stat` and `download_file` declarations noting the need for the impl to be thread safe
* As far as I can tell, all operations within the parallel block either do not modify state or operate on data that is intended for the specific package

Crzyrndm commented on src/vcpkg/binarycaching.cpp:
```diff
@@ -988,38 +993,36 @@ namespace
+                    }
+                    else
+                    {
+                        out_sink.println_warning(res.error());
```
Missed prints not being synchronised currently in previous assessment.
* https://github.com/microsoft/vcpkg-tool/pull/1290
* https://github.com/microsoft/vcpkg-tool/pull/1323

 This particular print could be extracted relatively easily and is the only print inside the parallel execution.

BillyONeal commented:
Thanks for your submission, I'm sorry at the time we didn't have the infrastructure to make this easy to do. We now have that so I've effectively implemented this for you in https://github.com/microsoft/vcpkg-tool/pull/2088 ; would you be willing to review over there? Thanks!
